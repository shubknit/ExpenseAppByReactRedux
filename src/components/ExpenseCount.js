import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { getVisibleExpenses } from '../selectors/expenses';

const ExpenseCount = (props) => {
	const {expenseList} = props;
	const getExpenseTotal = () => {
		let amount = 0 ;
	 	for( let i = 0 ; i< expenseList.length; i++){
	 		amount += expenseList[i].amount;
	 	}
	 	return amount;
	}

	return (
		<div className ="page-header">
			<div className = 'content-container'>
				<h1 className = "page-header__title"> Viewing { expenseList.length } expenses totalling ${ getExpenseTotal() }</h1>
				<Link to = '/create'>
						<button className = 'button'> Add Expense </button>
				</Link>
			</div>
		</div>
	)
}
const mapStateToProps = (state) => {
	return {
		expenseList: getVisibleExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseCount);