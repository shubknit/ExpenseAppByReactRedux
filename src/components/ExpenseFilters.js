import React from 'react';
import { Component } from 'react'; 
import { connect } from 'react-redux';
import { SetTextFilter, SetSortFilterByAmount, SetSortFilterByDate, SetStartDate, SetEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import 'react-dates/initialize';



export class ExpenseFilters extends Component {
	constructor(props){
		super(props);
		this.state = {
			startDate : this.props.filters.startDate,
			endDate: this.props.filters.endDate,
			focusedInput: null
		}
		this.filteredData = this.filteredData.bind(this);
		this.sortData = this.sortData.bind(this);
	}

	onDatesChange = ({startDate, endDate}) => {
		//this.setState({startDate, endDate});
		//this.props.dispatch(SetStartDate(startDate))
		this.props.onStartDateChange(startDate);
		//this.props.dispatch(SetEndDate(endDate))
		this.props.onEndDateChange(endDate);
	}

	onFocusChange = (focusedInput) => {
		this.setState({focusedInput})
	}

	filteredData(e){
		//this.props.dispatch(SetTextFilter(e.target.value));
		this.props.onTextChange(e.target.value);
	}

	sortData(e){
		let value = e.target.value;
		if(value === 'amount') {
			//this.props.dispatch(SetSortFilterByAmount(value));
			this.props.sortByAmount(value);
		}
		else if(value === 'date') {
			//this.props.dispatch(SetSortFilterByDate(value));
			this.props.sortByDate(value);
		}
	}

	render(){
		return(
			<div className="content-container">
			<div className="input-group">
				<div className="input-group__item">
					<input type="text" className="text-input"  placeholder="Search expenses"  onChange = {this.filteredData}/>
				</div>
				<div className="input-group__item">
					<select className="select" onChange = {this.sortData}>
						<option value="date">Date</option>
						<option value="amount">Amount</option>
					</select>
					
				</div>
				<div className="input-group__item">
						<DateRangePicker startDate = {this.state.startDate} endDate = {this.state.endDate}
						focusedInput = {this.state.focusedInput} onDatesChange = {this.onDatesChange}
						onFocusChange = {this.onFocusChange} startDateId = '1' endDateId = '1'
						isOutsideRange = {() => false} numberOfMonths = {1}
						showClearDates = {true}/>
				</div>
			</div>
		</div>
			)
	}
}

const mapStateToProps = (state) => {
	return {
		filters: state.filters
	}
}

const mapDispatchToProps = (dispatch) => {
	return {
		onStartDateChange: (startDate) => dispatch(SetStartDate(startDate)),
		onEndDateChange: (endDate) => dispatch(SetEndDate(endDate)),
		onTextChange: (text) => dispatch(SetTextFilter(text)),
		sortByAmount: (value) => dispatch(SetSortFilterByAmount()),
		sortByDate: (value) => dispatch(SetSortFilterByDate())
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilters);