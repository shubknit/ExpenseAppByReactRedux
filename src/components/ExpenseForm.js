import React from 'react';
import { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

export class ExpenseForm extends Component {
	constructor(props){
		super(props);
        this.onSubmit = this.onSubmit.bind(this);
        this.state = {
			description: props.formData ? props.formData.description : '',
			amount:  props.formData ? props.formData.amount : '',
			createdAt:  props.formData ? moment(props.formData.createdAt) : moment(),
			note:  props.formData ? props.formData.note : '',
			focused: false
		}
       
	}
	
	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({
			description
		}));	
	}

	onAmountChange = (e) => {
		const amount = e.target.value;
		this.setState(() => ({
			amount
		}))	
	}
	onNoteChange = (e) => {
		const note = e.target.value;
		this.setState(() => ({
			note
		}))	
	}

	onDateChange = (createdAt) => {
	    if (createdAt) {
	      this.setState(() => ({ createdAt }));
	    }
  	}

  	onFocusChange = ({focused}) => {
  		this.setState(() => ({
  			focused
  		}))
  	}

	onSubmit(e) {
		e.preventDefault();
		this.props.formDataSubmit({
			description: this.state.description,
			amount: parseInt(this.state.amount),
			createdAt: this.state.createdAt.valueOf(),
			note: this.state.note
		})
	}

	render(){
		return (
			<div>
				<form onSubmit = {this.onSubmit} className = 'form'>
					<label htmlFor = 'description'></label>
					<input type ="text" id ="description" placeholder ='description' className = 'text-input' 
					 autoFocus value = {this.state.description}
					onChange = {this.onDescriptionChange}/>

					<label htmlFor = 'amount'></label>
					<input type ="number" id ="amount" placeholder ='amount' className = 'text-input'
					value ={this.state.amount}   pattern="^\d*(\.\d{0,2})?$"
					onChange = {this.onAmountChange}/>

					<SingleDatePicker 
					  date={this.state.createdAt} isOutsideRange = {() => false} numberOfMonths = {1}
					  onDateChange={this.onDateChange} 
					  focused={this.state.focused} 
					  onFocusChange={this.onFocusChange} 
					/>
	
					<label htmlFor = 'note'></label>
					<textarea type ="text" id ="note" placeholder ='note' className = 'textarea'
					 required value = {this.state.note} onChange = {this.onNoteChange}/>
					<div><button  className = 'button' type ="submit"> Save Expense </button> </div>
				</form>
			</div>
		
		)
	}
}



