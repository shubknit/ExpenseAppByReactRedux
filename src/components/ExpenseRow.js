import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux'; 
import { startRemoveExpense } from '../actions/expenses';
import moment from 'moment';

export const ExpenseRow = (props) => {
	const {expenseRowData} = props;
	const date = moment(expenseRowData.createdAt).format('L');

	const remove = () => {
		//props.dispatch(RemoveExpense(expenseRowData.id));
		props.onRemove(expenseRowData.id);
	}
	return(
		<div className = 'list-item'>		
			<Link to = {`edit/${expenseRowData.id}`}>
				<div className ="">
					<div className = "col-1">
						<div> description : { expenseRowData.description } </div>
						<div> Date: { date } </div>
					</div>	
					<div className = "list">
						Amount: { expenseRowData.amount }
					</div>
					
				</div>
			</Link>
			<button className = 'button button--secondary' onClick = {remove}> Remove </button>
		</div>
	)
}

const mapDispatchToProps = (dispatch) => {
	return {
		onRemove: (expenseId) => dispatch(startRemoveExpense(expenseId))
	}
}
export default connect(undefined, mapDispatchToProps)(ExpenseRow);
