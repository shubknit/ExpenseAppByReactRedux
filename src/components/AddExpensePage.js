import React from 'react';
import { ExpenseHeader } from './ExpenseHeader';
import { ExpenseForm } from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export const AddExpensePage = (props) => {

	const submit = (data) => {
		//props.dispatch(AddExpense(data)); // To make it testable use mapDispatchToProps down below.
		props.onSubmit(data);
		props.history.push('/dashboard');	
	} 
	return(
		<div>
				<ExpenseHeader/>
				<div className ="page-header">
					<div className = 'content-container'>
						<h1 className = 'page-header__title'>Add Expense </h1>
					</div>
				</div>
				<div className = 'content-container'>
					<ExpenseForm formDataSubmit = {submit}/>
				</div>	

		</div>		
		)

}

const mapDispatchToProps = (dispatch) => {
	return {
		onSubmit: (expense) => dispatch(startAddExpense(expense))
	}
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
