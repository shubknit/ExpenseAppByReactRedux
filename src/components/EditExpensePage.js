import React from 'react';
import { ExpenseForm } from './ExpenseForm';
import { connect } from 'react-redux';
import { EditExpense } from '../actions/expenses';
import { startRemoveExpense, startEditExpense } from '../actions/expenses';

export const EditExpensePage = (props) => {
	const {expenseList} = props,
	 editExpenseId = props.match.params.id,
	 editExpenseFilteredData= expenseList.filter(data => data.id == editExpenseId),
	 formFilledData = editExpenseFilteredData[0],
	 { description, amount, note, createdAt } = editExpenseFilteredData[0];

	const submit = (data) => {
		//props.dispatch(EditExpense(editExpenseId,data));
		props.onEdit(editExpenseId, data);
		props.history.push('/dashboard');	
	} 

	const removeItem = () => {
		//props.dispatch(RemoveExpense(editExpenseId));
		props.onRemove(editExpenseId);
		props.history.push('/dashboard');
	}

	return(
		<div>
			<div className = 'page-header'>
				<div className ="content-container">
					<h1 className = 'page-header__title'>
						Edit Expense
					</h1>
				</div>
			</div>
			<div className = 'content-container'>
				<ExpenseForm formDataSubmit = {submit} formData = {formFilledData}/>
				<button className = 'button button--secondary' onClick = {removeItem}> Remove </button>
			</div>
		</div>	
		)

}
const mapStateToProps = (state) => {
	return {
		expenseList : state.expenses
	}
}

const mapDispatchToProps = (dispatch) => {
	return {
		onEdit: (expenseId, expense) => dispatch(startEditExpense(expenseId, expense)),
		onRemove: (expenseId) => dispatch(startRemoveExpense(expenseId))
	}
}
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
