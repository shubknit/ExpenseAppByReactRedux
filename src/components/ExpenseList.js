import ExpenseRow  from './ExpenseRow';
import { connect } from 'react-redux';
import { getVisibleExpenses } from '../selectors/expenses';

export const ExpenseList = ({expenses}) => 

		 (
		 	
		  <div className = "content-container">
				<div className = 'list-header'>
					<div> Expense</div>
					<div> Action</div>
				</div>
				<div className = 'list-body'>
					{ expenses.length > 0 ? 
					 expenses.map((data,i) => 
						<ExpenseRow key = {i} expenseRowData = {data}/>)
					 
					: <div className="list-item list-item--message"><span>No expenses</span></div>		
					
				}	
				</div>
			</div> 

		
		)



const mapStateToProps = (state) => {
	return {
		expenses: getVisibleExpenses(state.expenses, state.filters)
	}
};
export default connect(mapStateToProps)(ExpenseList)
		
	