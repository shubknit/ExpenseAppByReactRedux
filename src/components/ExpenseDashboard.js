import { Component } from 'react';
import { ExpenseHeader } from './ExpenseHeader';
import  ExpenseCount  from './ExpenseCount';
import ExpenseList from './ExpenseList';
import AddExpensePage from './AddExpensePage';
import EditExpensePage  from './EditExpensePage';
import  ExpenseFilters  from './ExpenseFilters';
//import uuid from 'uuid';


export class ExpenseDashboard extends Component {
	constructor() {
		super();
		this.saveState();
	
	}
	saveState(){
		localStorage.getItem('expenseData') ? 
		this.state = {
			expenseList : JSON.parse(localStorage.getItem('expenseData')) 
		} : this.state = {
			expenseList: [
				
			]
		}
	}
	setLocalStorage(data){
		localStorage.setItem('expenseData', JSON.stringify(data));
	}
	
	componentDidUpdate(){
		this.setLocalStorage(this.state.expenseList);
	}


	render(){
		return (
		   <div className = "main">
				<ExpenseHeader/>
				 <div className = 'main-dashboard'>
					<ExpenseCount/>
					<ExpenseFilters/>
					<ExpenseList /> 
		        </div>
		    </div>
		)

	}

}