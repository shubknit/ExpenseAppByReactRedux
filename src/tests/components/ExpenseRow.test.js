import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseRow } from '../../components/ExpenseRow';
import { expenseData } from '../expenseData';

test('should render ExpenseRow with expenses', () => {
  const wrapper = shallow(<ExpenseRow expenseRowData = {expenseData[0]}/>);
  expect(wrapper).toMatchSnapshot();
});


