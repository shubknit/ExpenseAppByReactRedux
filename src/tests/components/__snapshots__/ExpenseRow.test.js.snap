// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseRow with expenses 1`] = `
<div
  className="list-item"
>
  <Link
    replace={false}
    to="edit/1"
  >
    <div
      className=""
    >
      <div
        className="col-1"
      >
        <div>
           description : 
          rent
           
        </div>
        <div>
           Date: 
          01/01/1970
           
        </div>
      </div>
      <div
        className="list"
      >
        Amount: 
        100
      </div>
    </div>
  </Link>
  <button
    className="button button--secondary"
    onClick={[Function]}
  >
     Remove 
  </button>
</div>
`;
